---
title: "chromR"
author: "Brian J. Knaus"
date: "`r Sys.Date()`"
output: html_document
---


```{r global_options, include=FALSE}
#knitr::opts_chunk$set(fig.width=12, fig.height=8)
knitr::opts_chunk$set(fig.align='center')
```



In previous exercises we learned how to quality filter our VCF format GBS data.
Here we'll learn how to create chromR objects that can contain multiple forms of data.
In this exercise we'll use our variant data combined with reference sequence data.


First we'll import our VCF data into R.


```{r, results='hide'}
library(vcfR)
vcf <- read.vcfR('TASSEL_GBS0077_dp_filtered.vcf.gz')
```


Next, we'll read in our sequence data using the `read.dna()` function from the package ape.

```{r}
library(ape)
dna <- read.dna("data/Pr4671.fa", format="fasta")
```


## Data subset


The package vcfR was designed with teh intent to work on individual chromosomes.
Here we use the term 'chromosome' rather loosely.
Depending on the state of your reference your 'chromosomes' may actually be supercontigs or contigs.
By working on individual chromosomes it provides a natural manner to subset the large task of working on a genome into smaller tasks.
If you're not familiar with how to subset your data to individual chromosomes I've placed some documentation at [this link](https://github.com/knausb/vcfR_docs/blob/master/vignettes/subset_data_to_1chrom.Rmd).


First we'll explore the distribution of sequence lengths in the reference in order to find a good chromosome to use as an example.
Objects of class DNAbin can be stored as either matrices or lists, so we'll have to determine what we have before we extract this information.


```{r}
dna
```


The first line tells us that we have a list.
This means we can use the function `lapply()` to query our data structure.
For example.


```{r}
lapply(dna[1:4], length)
```


The return value of `lapply()` is a list, which is not compatible with our `hist()` function.
We can recast our list with `unlist()` so that we can plot a histogram.


```{r, fig.align='center', fig.width=10}
hist( unlist(lapply(dna, length)), col=8 )
names(dna)[unlist(lapply(dna, length)) > 120000]
```


We see that scaffold_1 is the longest scaffold.
It is convention to sort the sequences in a reference by length with the first sequence being the longest.
Its always good to check that this is the case with your genome.
Now we know that scaffold_1 is a good candidate to work on and we can subset the data to this scaffold.


```{r}
dna <- dna['scaffold_1']
```


When we move to the VCF data we observe a common issue.


```{r}
unique(vcf@fix[,'CHROM'])[1:10]
```


The names of the chromosomes in the VCF data are simply a number while the data in the sequence is more descriptive.
We can assume that the variant data from CHROM "1" corresponds with the sequence of scaffold_1.
But its important to reckognize that these are non-identical names that we need to keep track of.

```{r}
vcf <- vcf[grep("^1$", vcf@fix[,'CHROM']),]
```


Now that we have subset our data to scaffold_1 we can create an object of class chromR.


```{r}
chrom <- create.chromR(name="scaffold_1", vcf = vcf, seq = dna)
```


Note that the function warns us that the names are non-identical.
Because we've taken the time to make sure this is what we want to do we know we can safely proceed.
If you aren't sure of this, you should review.


```{r}
plot(chrom)
```


Note that 'DP' is no longer valid due to our previous manipulations.
Now that we've created an object of class chromR we update this fairly easily.
First, we'll extract the DP element from the gt region, then we'll repopulate the DP information with the rowSums from the DP matrix.


```{r}
dp <- extract.gt(chrom, element = "DP", as.numeric = TRUE)
chrom@var.info$DP <- rowSums(dp, na.rm = TRUE)
```


Now we can validate the change.


```{r}
plot(chrom)
```


I've tried to divide working with chromR objects into three phases: creation, processing and plotting.
Creation typically is dependent on funcitons and objects I did not create.
An example is our use of a DNAbin object created by `read.dna()`.
Because I did not create these functions or objects, there is little I can do about them if they do not perform as I would like.
Processing of a chromR object includes a sliding window analysis of GC content as well as summarizing variant incidence.
The functions that are a part of processing a chromR object are mostly my own.
This means I have some control as to how they perform.
Plotting of a chromR object depends largely on the base R graphics system, and we ask a lot of it.
Again, because this is not my code, I have little control over how it performs.



Now that we have created our chromR object, we can process it.


```{r}
chrom <- proc.chromR(chrom, win.size = 1e3)
```


The function `proc.chromR()` performs a sliding window analysis for variant incidence as well as GC content of the reference.
It also calculates some summary statistics.
For this example it performs quite well.
As the number of samples and variants increases it will present an increased computatinoal challenge that may require either patience or a strategy to subset to data.


Once we have an object of class chromR we can plot it with the funciton `chromoqc`.


```{r}
chromoqc(chrom)
```


The function `chromoqc()` is actually a wrapper to the function `chromo()`.
The latter function allows greater flexibility in plotting options.
It is also more complicated than `chromoqc()` which is really my attempt to guess how most users may want to parameterize `chromo()`.


One feature that may be rather subtle is the use of the elispes (...) as an argument in the function.
This is a way of saying 'other stuff too.'
The function `chromoqc` calls other functions to produce the plot and these functions all have parameters as well.
the elipses provide a way to pass parameters from `chromoqc()` to these other functions.
A notable option is the use of `xlim()`.
This option allows the user to focus in on a particular region of the chromosome.


```{r}
chromoqc(chrom, xlim=c(8e4, 10e4))
```


Use of chromR objects provides a way to integrate other forms of data with VCF data, such as sequence and annotation information.
The chromR object allows us to store exta information about VCF data while preserving the original data.
Lastly, the chromR object has plotiing capabilities to facilitate visualization of these data.
The use of chromR objects is intended to help researchers rapidly and easily scrutinize their data in order to provide the best possible data for downstream analyses.


